dev.py

```diff
@@ -35,47 +36,91 @@ def run_verbose(cmd_args, *args, **kwargs):
 
 
 def cmd_all_npm_build(args):
-    """"Build javascript code for all examples and templates"""
+    """Build javascript code for all examples and templates"""
     for project_dir in EXAMPLE_DIRECTORIES + TEMPLATE_DIRECTORIES:
         frontend_dir = next(project_dir.glob("*/frontend/"))
         run_verbose(["npm", "run", "build"], cwd=str(frontend_dir))
 
 
-def cmd_all_install_python_deps(args):
-    """"Install all dependencies needed to run e2e tests for all examples and templates"""
+def cmd_e2e_build_images(args):
+    """Build docker images for each component e2e tests"""
     for project_dir in EXAMPLE_DIRECTORIES + TEMPLATE_DIRECTORIES:
-        run_verbose(["pip", "install", "-e", ".[devel]"], cwd=str(project_dir))
+        e2e_dir = next(project_dir.glob("**/e2e/"), None)
+        if e2e_dir and os.listdir(e2e_dir):
+            # Define the image tag for the docker image
+            image_tag = (
+                f"component-template:py-{args.python_version}-st-{args.streamlit_version}-component-{project_dir.parts[-1]}"
+            )
+            # Build the docker image with specified build arguments
+            run_verbose(
+                [
+                    "docker",
+                    "build",
+                    ".",
+                    f"--build-arg=STREAMLIT_VERSION={args.streamlit_version}",
+                    f"--build-arg=PYTHON_VERSION={args.python_version}",
+                    f"--tag={image_tag}",
+                    "--progress=plain",
+                ],
+                env={**os.environ, "DOCKER_BUILDKIT": "1"},
+            )
 
 
-def cmd_all_install_wheel_packages(args):
-    """"Install wheel packages of all examples and templates for e2e tests"""
+def cmd_e2e_run(args):
+    """Run e2e tests for all examples and templates in separate docker images"""
     for project_dir in EXAMPLE_DIRECTORIES + TEMPLATE_DIRECTORIES:
-        wheel_files = list(project_dir.glob("dist/*.whl"))
-        if wheel_files:
-            wheel_file = wheel_files[0]
-            run_verbose(["pip", "install", str(wheel_file)], cwd=str(project_dir))
-        else:
-            print(f"No wheel files found in {project_dir}")
-
-
-def cmd_install_browsers(args):
-    """"Install multiple browsers to run e2e for all examples and templates"""
-    run_verbose(["playwright", "install", "webkit", "chromium", "firefox", "--with-deps"])
-
-
-def cmd_all_run_e2e(args):
-    """"Run e2e tests for all examples and templates"""
+        container_name = project_dir.parts[-1]
+        image_tag = (
+            f"component-template:py-{args.python_version}-st-{args.streamlit_version}-component-{container_name}"
+        )
+        e2e_dir = next(project_dir.glob("**/e2e/"), None)
+        if e2e_dir and os.listdir(e2e_dir):
+            mount_catalog = e2e_dir.parts[-2]
+            run_verbose([
+                "docker",
+                "run",
+                "--tty",
```

Comment:
```suggestion
                "--tty",
                "--rm",
```
Running containers with `--rm` flag is good for those containers that you use for very short while just to accomplish something, e.g., compile your application inside a container, or just testing something that it works, and then you are know it's a short lived container and you tell your Docker daemon that once it's done running, erase everything related to it and save the disk space.
https://stackoverflow.com/questions/49726272/what-is-the-rm-flag-doing